python/pdo/contract/request.py

```diff
@@ -159,7 +159,15 @@ def evaluate(self) :
 
         try :
             if response_parsed['Status'] and response_parsed['StateChanged'] :
-                contract_response = UpdateStateResponse(self, response_parsed)
+                # the contract reported that the state changed, lets check the state
+                # hashes to make sure that it really did change
+                if response_parsed['StateHash'] == self.contract_state.get_state_hash('b64') :
+                    # Note that changing the value of StateChanged should not impact the
+                    # verification of the enclave signature; still need to verify this
+                    response_parsed['StateChanged'] = False
```

Comment:
Not sure I understand this.
The contract said that the state has changed. Here the code is checking that claim. But if the state has not changed (which should not be the case at this point; rather this should be handled in the else block at line 171), it updates the StateChanged flag which comes from the contract?!

Reply:
The contract said the state changed (we put in this flag as an optimization so the contract could say that the state didn't change). It didn't actually change. So the flag is wrong. We fix the flag here. Not sure what 171 has to do with it?

Reply:
Why does the contract say something "wrong"?
This is not a test, and the contract knows when the state changes -- thanks to our outstanding KVS!

Reply:
Because we don't even attempt to return a new state unless the contract suggests that it might have changed the state. This flag is more "this method *might* change state".

Reply:
On a slightly different note, i think, we also need to consider (not in this PR) the implications of "what if the state actually changed, but the contract says state not changed"? As long as the commits are done by the client, there is no issue, but in the past we have talked by the eservice submitting a txn batch directly on behalf of the clients. If the txn submission is outsourced, such discrepancies would be an issue.

Reply:
agreed that the return value needs to be re-considered. i know why we added this long ago, but not sure we need it any more. this is, however, part of a much bigger discussion that will require ledger changes (e.g. contract-specified transaction metadata)